### ISSUE TITLE
Introduce a Ballerina connector for AWS Secret Manager

### ISSUE THREAD

**Issue by ayeshLK** [ballerina-platform/ballerina-library#7471]
### Problem

Currently there is a POC in-progress with a customer which requires a connector for AWS Secret Manager and Ballerina does not have a connector for AWS secret manager

### Proposed Solution

Introduce a Ballerina connector for AWS Secret Manger

**Comment by TharmiganK** [comment]
Find the comments below:
1. Are the field names for the types correct? I think the actual field name in the response and request are in upper case: https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_DescribeSecret.html#API_DescribeSecret_Examples

2. Do we need the common `Version` parameter in the design - https://docs.aws.amazon.com/secretsmanager/latest/apireference/CommonParameters.html

3. The `time:Utc` is a tuple type, better to have the timestamp values as decimal

4. Better to define this as a enum: 
    ```ballerina
    public type StagingStatus "AWSCURRENT"|"AWSPENDING"|"AWSPREVIOUS";
    ```

5. The errors can be sub-typed right? https://docs.aws.amazon.com/secretsmanager/latest/apireference/CommonErrors.html

**Comment by ayeshLK** [comment]
>     1. Are the field names for the types correct? I think the actual field name in the response and request are in upper case: https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_DescribeSecret.html#API_DescribeSecret_Examples

This is intentional as in Ballerina we use camel-case for naming.
 
>     2. Do we need the common `Version` parameter in the design - https://docs.aws.amazon.com/secretsmanager/latest/apireference/CommonParameters.html

No need as these values are internally handled by the Java SDK.
 
>     3. The `time:Utc` is a tuple type, better to have the timestamp values as decimal

We can do that, but I thought `time:Utc` was our preferred approach. @daneshk WDYT ?

>     4. Better to define this as a enum:
>        ```ballerina
>        public type StagingStatus "AWSCURRENT"|"AWSPENDING"|"AWSPREVIOUS";
>        ```

+1

>     5. The errors can be sub-typed right? https://docs.aws.amazon.com/secretsmanager/latest/apireference/CommonErrors.html

Yes, but since this is the initial version and we have a strict deadline to deliver, we can do this later.

**Comment by daneshk** [comment]
> > ```
> > 1. Are the field names for the types correct? I think the actual field name in the response and request are in upper case: https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_DescribeSecret.html#API_DescribeSecret_Examples
> > ```
> 
> This is intentional as in Ballerina we use camel-case for naming.
> 
> > ```
> > 2. Do we need the common `Version` parameter in the design - https://docs.aws.amazon.com/secretsmanager/latest/apireference/CommonParameters.html
> > ```
> 
> No need as these values are internally handled by the Java SDK.
> 
> > ```
> > 3. The `time:Utc` is a tuple type, better to have the timestamp values as decimal
> > ```
> 
> We can do that, but I thought `time:Utc` was our preferred approach. @daneshk WDYT ?

Yes, using `time:Utc` is our preferred approach. As we are using Java SDK underneath, we can convert from a decimal value.

**Comment by ayeshLK** [comment]
Currently, as per our initial design, retrieving a secret from AWS Secrets Manager requires users to provide the `secretId` as a named parameter, as shown below:

```ballerina
secretmanager:SecretValue secret = check secretManager->getSecretValue(secretId = "<secret-id>");
```

While this design works, `secretId` is a mandatory parameter for this operation. To improve the usability and streamline the API, we can modify the design to allow users to pass the `secretId` as a positional argument when the operation only requires the `secretId`. This change simplifies the usage and avoids the need for the named-parameter syntax in such cases.

```ballerina
// client API
remote function getSecretValue(SecretId secretId, *GetSecretValueRequest request) returns SecretValue|Error;

// `GetSecretValueRequest` type
public type GetSecretValueRequest record {|
    @constraint:String {
        minLength: 32,
        maxLength: 64
    }
    string versionId?;
    @constraint:String {
        minLength: 1,
        maxLength: 256
    }
    string versionStage?;
|};

// `SecretId` type
@constraint:String {
    minLength: 1,
    maxLength: 2048
}
public type SecretId string;
```

With this suggestion user can use the client like this:
```ballerina
secretmanager:SecretValue secret = check secretManager->getSecretValue("<secret-id>");
```

@daneshk @ThisaruGuruge WDYT ?

**Comment by daneshk** [comment]
I am +1 to pass the secretId as a positional argument. Is this okay to group `versionId` and `version stage` only as named parameters?

**Comment by ayeshLK** [comment]
> I am +1 to pass the secretId as a positional argument. Is this okay to group `versionId` and `version stage` only as named parameters?

@daneshk we can also have those as positional arguments, but since these are optional values the API would look like this:

```ballerina
remote function getSecretValue(SecretId secretId, string? versionId = (), string? versionStage = ()) returns SecretValue|Error;
```

**Comment by ayeshLK** [comment]
> Currently, as per our initial design, retrieving a secret from AWS Secrets Manager requires users to provide the `secretId` as a named parameter, as shown below:
> 
> ```ballerina
> secretmanager:SecretValue secret = check secretManager->getSecretValue(secretId = "<secret-id>");
> ```
> 
> While this design works, `secretId` is a mandatory parameter for this operation. To improve the usability and streamline the API, we can modify the design to allow users to pass the `secretId` as a positional argument when the operation only requires the `secretId`. This change simplifies the usage and avoids the need for the named-parameter syntax in such cases.
> 
> ```ballerina
> // client API
> remote function getSecretValue(SecretId secretId, *GetSecretValueRequest request) returns SecretValue|Error;
> 
> // `GetSecretValueRequest` type
> public type GetSecretValueRequest record {|
>     @constraint:String {
>         minLength: 32,
>         maxLength: 64
>     }
>     string versionId?;
>     @constraint:String {
>         minLength: 1,
>         maxLength: 256
>     }
>     string versionStage?;
> |};
> 
> // `SecretId` type
> @constraint:String {
>     minLength: 1,
>     maxLength: 2048
> }
> public type SecretId string;
> ```
> 
> With this suggestion user can use the client like this:
> 
> ```ballerina
> secretmanager:SecretValue secret = check secretManager->getSecretValue("<secret-id>");
> ```
> 
> @daneshk @ThisaruGuruge WDYT ?

With the suggested change `GetSecretValueRequest` might not be appropriate as it only contains secret-version related request parameters. We will rename the type as `SecretVersionSelector`.

**Comment by ayeshLK** [comment]
Completed with the following PR:
- https://github.com/ballerina-platform/module-ballerinax-aws.secretmanager/pull/2
